ffs/minerselector/reptop/reptop.go

```diff
@@ -32,30 +45,59 @@ func (rt *RepTop) GetMiners(n int, f ffs.MinerSelectorFilter) ([]ffs.MinerPropos
 	if n < 1 {
 		return nil, fmt.Errorf("the number of miners should be greater than zero")
 	}
-	ms, err := rt.rm.QueryMiners(f.ExcludedMiners, f.CountryCodes, f.TrustedMiners)
+
+	// We start directly targeting trusted miners without relying on reputation index.
+	// This is done to circumvent index building restrictions such as excluding miners
+	// with zero power. Trusted miners are trusted by definition, so if they have zero
+	// power, that's a risk that the client is accepting.
+	trustedMiners := rt.genTrustedMiners(f, n)
+
+	// The remaining needed miners are gathered from the reputation index.
+	reputationMiners, err := rt.genFromReputation(f, n-len(trustedMiners))
+	if err != nil {
+		return nil, fmt.Errorf("getting miners from reputation module: %s", err)
+	}
+
+	// Return both lists.
+	return append(trustedMiners, reputationMiners...), nil
+}
+
+func (rt *RepTop) genTrustedMiners(f ffs.MinerSelectorFilter, n int) []ffs.MinerProposal {
+	ret := make([]ffs.MinerProposal, 0, len(f.TrustedMiners))
+	for _, m := range f.TrustedMiners {
+		mp, err := rt.getMinerProposal(f, m)
+		if err != nil {
+			log.Warnf("trusted miner query asking: %s", err)
+			continue
+		}
+		ret = append(ret, mp)
+		if len(ret) == n {
+			break
+		}
+	}
+
+	return ret
+}
+
+func (rt *RepTop) genFromReputation(f ffs.MinerSelectorFilter, n int) ([]ffs.MinerProposal, error) {
+	if n == 0 {
+		return nil, nil
+	}
+	ms, err := rt.rm.QueryMiners(f.TrustedMiners, f.CountryCodes, nil)
 	if err != nil {
 		return nil, fmt.Errorf("getting miners from reputation module: %s", err)
 	}
 	if len(ms) < n {
-		return nil, fmt.Errorf("not enough miners that satisfy the constraints")
+		return nil, fmt.Errorf("not enough miners from reputation module to satisfy the constraints")
 	}
-	aidx := rt.ai.Get()
+
 	res := make([]ffs.MinerProposal, 0, n)
 	for _, m := range ms {
-		sa, ok := aidx.Storage[m.Addr]
-		if !ok {
-			continue
-		}
-		if f.MaxPrice > 0 && sa.Price > f.MaxPrice {
-			continue
+		mp, err := rt.getMinerProposal(f, m.Addr)
+		if err != nil {
+			continue // Cached miner isn't replying to query-ask now, skip.
 		}
-		if f.PieceSize < sa.MinPieceSize || f.PieceSize > sa.MaxPieceSize {
-			continue
-		}
-		res = append(res, ffs.MinerProposal{
-			Addr:       sa.Miner,
-			EpochPrice: sa.Price,
-		})
+		res = append(res, mp)
 		if len(res) == n {
 			break
 		}
```

Comment:
Ugly diff from GH, but the improtant part reg a question you had yesterday @asutula : 
- `rt.rm.Queryminers` returns all the list, it isn't capped.
- The `for` loop only early-break if `len(res)==n`

So that basically means that if potential candidates are discarded by a fresh query-ask, then the loop continues with the next.